c/driver/postgresql/connection.cc

```diff
@@ -47,7 +48,97 @@ AdbcStatusCode PostgresConnection::GetTableSchema(const char* catalog,
                                                   const char* table_name,
                                                   struct ArrowSchema* schema,
                                                   struct AdbcError* error) {
-  return ADBC_STATUS_NOT_IMPLEMENTED;
+  // TODO: sqlite uses a StringBuilder class that seems roughly equivalent
+  // to a similar tool in arrow/util/string_builder.h but wasn't clear on
+  // structure and how to use, so relying on simplistic appends for now
+  AdbcStatusCode final_status;
+
+  std::string query =
+      "SELECT attname, atttypid "
+      "FROM pg_catalog.pg_class AS cls "
+      "INNER JOIN pg_catalog.pg_attribute AS attr ON cls.oid = attr.attrelid "
+      "INNER JOIN pg_catalog.pg_type AS typ ON attr.atttypid = typ.oid "
+      "WHERE attr.attnum >= 0 AND cls.oid = '";
+  if (db_schema != nullptr) {
+    query.append(db_schema);
+    query.append(".");
+  }
+  query.append(table_name);
+  query.append("'::regclass::oid");
+
+  // char* stmt = PQescapeLiteral(conn_, query.c_str(), query.length());
+  // if (stmt == nullptr) {
+  //   SetError(error, "Failed to get table schema: ", PQerrorMessage(conn_));
+  //  return ADBC_STATUS_INVALID_ARGUMENT;
+  // }
+
+  pg_result* result = PQexec(conn_, query.c_str());
+  // PQfreemem(stmt);
+
+  ExecStatusType pq_status = PQresultStatus(result);
+  if (pq_status == PGRES_TUPLES_OK) {
+    int num_rows = PQntuples(result);
+    ArrowSchemaInit(schema);
+    CHECK_NA_ADBC(ArrowSchemaSetTypeStruct(schema, num_rows), error);
+
+    // TODO: much of this code is copied from statement.cc InferSchema
```

Comment:
I _think_ this should be moved out of statement.cc and probably put into type.cc